helios-core/src/main/kotlin/helios/instances/instancesJava.kt

```diff
@@ -127,3 +128,30 @@ interface MapDecoderInstance<A, B> : Decoder<Map<A, B>> {
   }
 
 }
+
+interface EnumEncoderInstance<E : Enum<E>> : Encoder<Enum<E>> {
+
+  override fun Enum<E>.encode(): Json = JsString(name)
+
+  companion object {
+    operator fun <E : Enum<E>> invoke(): Encoder<Enum<E>> =
+      object : EnumEncoderInstance<E> {}
+  }
+}
+
+interface EnumDecoderInstance<E : Enum<E>> : Decoder<Enum<E>> {
+
+  fun enumClass(): KClass<E>
+
+  override fun decode(value: Json): Either<DecodingError, Enum<E>> =
+    value.asJsString()
+      .toEither { StringDecodingError(value) }
+      .map { java.lang.Enum.valueOf<E>(enumClass().java, it.value.toString()) }
```

Comment:
Unfortunately I have to use `java.lang.Enum.valueOf` here because Kotlin's `Enum` does not declare a `valueOf`, even though the compiler emits bytecode which simply calls `java.lang.Enum.valueOf`.

Reply:
Wouldn't that throw in case of error? try/catch it on a flatMap instead :)

Reply:
Yes, you are right. I added two tests for that. It won't throw an NPE, though, because Kotlin guarantees neither of the two preconditions for an NPE will be null.